clippy_lints/src/derive.rs

```diff
@@ -419,3 +453,43 @@ impl<'tcx> Visitor<'tcx> for UnsafeVisitor<'_, 'tcx> {
         self.cx.tcx.hir()
     }
 }
+
+/// Implementation of the `DERIVE_PARTIAL_EQ_WITHOUT_EQ` lint.
+fn check_partial_eq_without_eq<'tcx>(
+    cx: &LateContext<'tcx>,
+    span: Span,
+    trait_ref: &TraitRef<'_>,
+    ty: Ty<'tcx>,
+    peq_is_automatically_derived: bool,
+) {
+    if_chain! {
+        if let ty::Adt(adt, substs) = ty.kind();
+        if peq_is_automatically_derived;
+        if let Some(eq_trait_def_id) = cx.tcx.get_diagnostic_item(sym::Eq);
+        if let Some(def_id) = trait_ref.trait_def_id();
+        if cx.tcx.is_diagnostic_item(sym::PartialEq, def_id);
+        then {
+            if !implements_trait(cx, ty, eq_trait_def_id, substs) {
+                // If all of our fields implement `Eq`, we can implement `Eq` too
+                for variant in adt.variants() {
+                    for field in &variant.fields {
+                        let ty = field.ty(cx.tcx, substs);
+
+                        if !implements_trait(cx, ty, eq_trait_def_id, substs) {
+                            return;
+                        }
+                    }
+                }
+
+                span_lint_and_help(
```

Comment:
I tried making this work with a suggestion like this:

```rust
span_lint_and_sugg(
    cx,
    DERIVE_PARTIAL_EQ_WITHOUT_EQ,
    span,
    "you are deriving `PartialEq` and can implement `Eq`",
    "consider deriving `Eq` as well",
    "PartialEq, Eq".to_string(),
    Applicability::MachineApplicable,
)
```

But when I ran this it just wouldn't print a suggestion at all, not sure what I did wrong.

Reply:
What did it print? That should have worked.

Reply:
So when I switch to `span_lint_and_sugg` on my branch, here's the stderr diff `cargo uitest` gives me: https://pastebin.com/v7Eg5kDp

Note that the only difference is that the ``help: consider deriving `Eq` as well`` is dropped.

Here's the JSON that `cargo uitest` spat out for one of the errors: https://pastebin.com/Gr6w967J

Looks like `suggested_replacement` and `suggestion_applicability` are null?

Reply:
It's because the span is for the impl block generated by the derive. You need to use `span.ctxt().outer_expn_data().call_site`.

Reply:
Ahhh, that makes a lot of sense, thanks! Fixed in the latest revision.